### PR TITLE
Allow select correct invoice -> invoice_number can be "duplicated" (every year)

### PR DESCRIPTION
**Pull request by ShaiMagal**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.2.x
| Description?      | OrderInvoice::getInvoiceByNumber can return more records, but ->getValue return only one. If you are resetting invoice number every year, you will have in DB table "ps_order_invoice" multiple invoices with same number (check screenshot with real example). So we need make small improvement for function getInvoiceByNumber to add "id_order" to WHERE in sql command for easily access for module developers or other connecting services.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | 
| UI Tests          | 
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | https://www.openservis.cz/

Untitled

### CODE REVIEW

**Review by matthieu-rolland — APPROVED** [review]
seems like a nice useful improvement 👍 

You should probably create an issue for this and link your PR

**Review by matthieu-rolland — DISMISSED** [review]
Ah wait, shouldn't this target develop since it's not a regression ?

**Comment by Hlavtox** [comment]
@matthieu-rolland Well, it's a bit of a bug, it just randomly returns an invoice from random year :D

File: classes/order/OrderInvoice.php
Comment (Hlavtox) [inline classes/order/OrderInvoice.php]:
```suggestion
    /**
     * Returns OrderInvoice for a specific invoice number and order ID. It's highly recommended to also provide an order ID, because you may end up with a different invoice than you wanted.
     *
     * @param string|int $id_invoice
     * @param int $id_order
     * 
     * @return OrderInvoice|false
     */
    public static function getInvoiceByNumber($id_invoice, $id_order = null)
```

File: classes/order/OrderInvoice.php
Comment (Hlavtox) [inline classes/order/OrderInvoice.php]:
```suggestion
            WHERE `number` = ' . (int) $id_invoice . (!empty($id_order) ? ' && `id_order` = ' . (int) $id_order : '')
```

**Comment by ShaiMagal** [comment]
@matthieu-rolland It isn't BC, so I think it isn't necessary to do for develop I think. I think, this should be in minor version.

Like already wrote @Hlavtox "Well, it's a bit of a bug, it just randomly returns an invoice from random year :D" -> tottally true...

We should make this fix for 8.x, module creators can "benefit" with this fix sooner, than develop I think :)

**Comment by ShaiMagal** [comment]
@matthieu-rolland @Hlavtox guys?

**Comment by ShaiMagal** [comment]
@matthieu-rolland @Hlavtox ping

**Comment by ShaiMagal** [comment]
@matthieu-rolland @Hlavtox ping again

**Comment by ShaiMagal** [comment]
@matthieu-rolland ping

**Review by Hlavtox — COMMENTED** [review]
@ShaiMagal Bro, looking at the PR once more.

I think we can improve it a bit, then it will be okay and we can merge it. :-)

Please, can you change the variable `$id_invoice` name to `$invoiceNumber`? The parameter name being `$id_invoice` is very confusing and it makes it look like some unique ID, which is not.

**Comment by ShaiMagal** [comment]
@Hlavtox Done, but it's not related to my PR, it's something else :-D But OK!

**Comment by ShaiMagal** [comment]
@Hlavtox @matthieu-rolland @kpodemski Guys, what about this PR?

**Review by Hlavtox — DISMISSED** [review]

**Comment by ShaiMagal** [comment]
@matthieu-rolland ?

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.1.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))
* The `How to test` shouldn&#039;t be empty. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#how-to-test))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

**Comment by matks** [comment]
Hi, I have changed the target branch to 8.2.x because of https://github.com/PrestaShop/PrestaShop/issues/36758

**Comment by ShaiMagal** [comment]
@matks Thank you.

But @matthieu-rolland is blocking this PR :)

**Review by Hlavtox — DISMISSED** [review]

**Comment by ShaiMagal** [comment]
@matks can we approve this?

File: classes/order/OrderInvoice.php
Comment (kpodemski) [inline classes/order/OrderInvoice.php]:
```suggestion
    public static function getInvoiceByNumber($invoiceNumber, $orderId = null)
```

File: classes/order/OrderInvoice.php
Comment (kpodemski) [inline classes/order/OrderInvoice.php]:
```suggestion
            WHERE `number` = ' . (int) $invoiceNumber . (!empty($orderId) ? ' AND `id_order` = ' . (int) $orderId : '')
```

**Review by kpodemski — COMMENTED** [review]
I think it makes sense. I added to little suggestions.

**Comment by ShaiMagal** [comment]
@kpodemski I commited suggestions :)

**Review by Hlavtox — DISMISSED** [review]

**Comment by Hlavtox** [comment]
@ShaiMagal I rebased it, fixed a tiny issue with PHPDoc, add one more comment and merged all commints. Ready to go. :-)

**Review by Hlavtox — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Hello @ShaiMagal ,

Tested the following :

- Made multiple orders > Changed status to Payment accepted
- Checked the database : number of invoice is always different
- Changed different options in Orders > Invoices >  `Reset sequential invoice number at the beginning of the year` : YES &  `Add current year to invoice number` & `Invoice number`

Although, as I didn't change year for my test, the test couldn't be completed properly. 

As the automated testes and partial functional  testing is OK : This is QA ✅

**Comment by ps-jarvis** [comment]
PR merged, well done!

Message to @PrestaShop/committers: do not forget to milestone it before the merge.

**Comment by Hlavtox** [comment]
Thank you @florine2623 :-)

**Pull request merged**
